pkg/controller/master/addon/common.go

```diff
@@ -0,0 +1,402 @@
+			addonOwned = true
+			break
+		}
+	}
+	return hc, addonOwned, nil
+}
+
+// check if update is needed, when needed, also return values related string for further use
+func (h *Handler) isHelmchartUpdateNeeded(aObj *harvesterv1.Addon, hc *helmv1.HelmChart) (bool, string, error) {
+	vals, err := defaultValues(aObj)
+	if err != nil {
+		return false, "", fmt.Errorf("error generating default values of addon %s/%s: %v", aObj.Namespace, aObj.Name, err)
+	}
+
+	return (hc.Spec.ValuesContent != vals || hc.Spec.Version != aObj.Spec.Version || hc.Spec.Chart != aObj.Spec.Chart || hc.Spec.Repo != aObj.Spec.Repo), vals, nil
+}
+
+// deploy a new chart
+func (h *Handler) deployHelmChart(aObj *harvesterv1.Addon) error {
+	vals, err := defaultValues(aObj)
+	if err != nil {
+		return err
+	}
+
+	hc := &helmv1.HelmChart{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      aObj.Name,
+			Namespace: aObj.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: aObj.APIVersion,
+					Kind:       aObj.Kind,
+					Name:       aObj.Name,
+					UID:        aObj.UID,
+				},
+			},
+		},
+		Spec: helmv1.HelmChartSpec{
+			Chart:         aObj.Spec.Chart,
+			Repo:          aObj.Spec.Repo,
+			ValuesContent: vals,
+			Version:       aObj.Spec.Version,
+		},
+	}
+	_, err = h.helm.Create(hc)
+	if err != nil {
+		return fmt.Errorf("error creating helmchart object %v", err)
+	}
+
+	return nil
+}
+
+func defaultValues(a *harvesterv1.Addon) (string, error) {
+	if a.Spec.ValuesContent != "" {
+		return a.Spec.ValuesContent, nil
+	}
+
+	valsEncoded, ok := a.Annotations[util.AddonValuesAnnotation]
+	if ok {
+		valByte, err := base64.StdEncoding.DecodeString(valsEncoded)
+		if err != nil {
+			return "", fmt.Errorf("error decoding addon defaults: %v", err)
+		}
+
+		return string(valByte), nil
+	}
+	// no overrides. Use packaged chart defaults
+	return "", nil
+}
+
+func isJobComplete(j *batchv1.Job) bool {
+	if j.Status.CompletionTime != nil {
+		return true
+	}
+
+	for _, v := range j.Status.Conditions {
+		if v.Type == batchv1.JobFailed && v.Reason == "BackoffLimitExceeded" {
+			return true
+		}
+	}
+
+	return false
+}
+
+// check if timeout; user can adjust it via annotations AnnotationAddonOperationTimeout
+// return bool: if timeout; bool: if need to update the LastTransitionTime
+func (h *Handler) isAddonOperationTimeout(aObj *harvesterv1.Addon, cond condition.Cond) (bool, bool) {
+	start, _ := h.getAddonConditionLastTransitionTime(aObj, cond)
+	// no timestamp, or parse error, need re-update
+	if start == nil {
+		return false, true
+	}
+
+	timeout := time.Duration(addonTimeout)
+	if aObj.Annotations != nil {
+		if timeoutAnno, ok := aObj.Annotations[util.AnnotationAddonOperationTimeout]; ok {
+			i, err := strconv.ParseInt(timeoutAnno, 10, 32)
+			if err != nil {
+				logrus.Debugf("addon %s has invalid AnnotationAddonOperationTimeout, fallback to default value %v", aObj.Name, timeout)
+			} else if i >= addonTimeoutMin && i <= addonTimeoutMax {
```

Comment:
let validator do an check is OK, it could be added later.

the controller can tolerent the invalid value, as we have a default value to fallback.

Reply:
em, the timeout should be added to document rather than validator the annotations. in k8s, annotations are flexible to add/modify, it only works when annotations meets declare, but rarely using validator to validate it.  the addon may be added from bootstrap stage, or upgraded from previous;  validator to force check, may break something.